File: model_1/CS3233_2024_Midterm/endangered_safari_zone_1_solution.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <cmath>

using namespace std;

const int MAXN = 100005;

struct Query {
    int time, l, r, idx;
};

struct Update {
    int pos, old_val, new_val;
};

int block_size;
vector<int> arr;
vector<Query> queries;
vector<Update> updates;
vector<int> ans;
vector<int> freq;
int distinct_count, bad_count;
int N, M, K;

bool cmp(const Query &a, const Query &b) {
    int block_a_l = a.l / block_size;
    int block_b_l = b.l / block_size;
    if (block_a_l != block_b_l) {
        return block_a_l < block_b_l;
    }
    int block_a_r = a.r / block_size;
    int block_b_r = b.r / block_size;
    if (block_a_r != block_b_r) {
        if (block_a_l % 2 == 0) {
            return block_a_r < block_b_r;
        } else {
            return block_a_r > block_b_r;
        }
    }
    return a.time < b.time;
}

void add_element(int i) {
    int species = arr[i];
    int &count = freq[species];
    if (count == 0) {
        distinct_count++;
    }
    count++;
    if (count == K + 1) {
        bad_count++;
    }
}

void remove_element(int i) {
    int species = arr[i];
    int &count = freq[species];
    if (count == K + 1) {
        bad_count--;
    }
    count--;
    if (count == 0) {
        distinct_count--;
    }
}

void apply_update(const Update &u, int current_l, int current_r) {
    if (u.old_val == u.new_val) {
        return;
    }
    if (current_l <= u.pos && u.pos <= current_r) {
        remove_element(u.pos);
        arr[u.pos] = u.new_val;
        add_element(u.pos);
    } else {
        arr[u.pos] = u.new_val;
    }
}

void unapply_update(const Update &u, int current_l, int current_r) {
    if (u.old_val == u.new_val) {
        return;
    }
    if (current_l <= u.pos && u.pos <= current_r) {
        remove_element(u.pos);
        arr[u.pos] = u.old_val;
        add_element(u.pos);
    } else {
        arr[u.pos] = u.old_val;
    }
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    cin >> N >> M >> K;
    arr.resize(N);
    for (int i = 0; i < N; i++) {
        cin >> arr[i];
        arr[i]--;
    }

    vector<int> current_array = arr;
    int time_count = 0;
    int query_count = 0;
    vector<int> output_ans;

    for (int i = 0; i < M; i++) {
        int q, a, b;
        cin >> q >> a >> b;
        if (q == 1) {
            a--;
            int b0 = b - 1;
            int old_val = current_array[a];
            if (old_val != b0) {
                current_array[a] = b0;
            }
            updates.push_back({a, old_val, b0});
            time_count++;
        } else {
            a--;
            b--;
            queries.push_back({time_count, a, b, query_count++});
        }
    }

    if (queries.empty()) {
        return 0;
    }

    block_size = (int)pow(N, 2.0/3.0);
    if (block_size < 1) block_size = 1;
    sort(queries.begin(), queries.end(), cmp);

    arr = current_array;
    freq.assign(MAXN, 0);
    distinct_count = 0;
    bad_count = 0;
    int current_l = 0, current_r = -1, current_time = updates.size();
    ans.resize(query_count);

    for (const auto &q : queries) {
        while (current_time < q.time) {
            apply_update(updates[current_time], current_l, current_r);
            current_time++;
        }
        while (current_time > q.time) {
            current_time--;
            unapply_update(updates[current_time], current_l, current_r);
        }

        while (current_l > q.l) {
            add_element(--current_l);
        }
        while (current_r < q.r) {
            add_element(++current_r);
        }
        while (current_l < q.l) {
            remove_element(current_l++);
        }
        while (current_r > q.r) {
            remove_element(current_r--);
        }

        ans[q.idx] = distinct_count - bad_count;
    }

    for (int i = 0; i < query_count; i++) {
        cout << ans[i] << '\n';
    }

    return 0;
}

File: model_1/CS3233_2024_Midterm/endangered_safari_zone_1_solution.py
import sys
import math

def main():
    data = sys.stdin.read().split()
    if not data:
        return
    
    it = iter(data)
    n = int(next(it)); m_val = int(next(it)); K_val = int(next(it))
    arr0 = [int(next(it)) - 1 for _ in range(n)]
    
    queries = []
    updates = []
    current_arr = arr0[:]
    for i in range(m_val):
        q_type = int(next(it)); a = int(next(it)); b = int(next(it))
        if q_type == 1:
            pos = a - 1
            new_val = b - 1
            old_val = current_arr[pos]
            updates.append((pos, old_val, new_val))
            current_arr[pos] = new_val
        else:
            l = a - 1; r = b - 1
            time_stamp = len(updates)
            queries.append((time_stamp, l, r, len(queries)))
    
    if not queries:
        return
    
    block_size = int(round(n ** (2.0/3.0)))
    if block_size < 1:
        block_size = 1
    
    def mo_cmp(query):
        time, l, r, idx = query
        l_block = l // block_size
        r_block = r // block_size
        return (l_block, r_block if l_block % 2 == 0 else -r_block, time)
    
    queries_sorted = sorted(queries, key=mo_cmp)
    
    arr = current_arr[:]
    freq = [0] * (n + 1)
    distinct_count = 0
    bad_count = 0
    
    def add_element(i):
        nonlocal distinct_count, bad_count
        x = arr[i]
        if freq[x] == 0:
            distinct_count += 1
        freq[x] += 1
        if freq[x] == K_val + 1:
            bad_count += 1
    
    def remove_element(i):
        nonlocal distinct_count, bad_count
        x = arr[i]
        if freq[x] == K_val + 1:
            bad_count -= 1
        freq[x] -= 1
        if freq[x] == 0:
            distinct_count -= 1
    
    def apply_update(update, current_l, current_r):
        nonlocal arr
        pos, old_val, new_val = update
        if current_l <= pos <= current_r:
            remove_element(pos)
        arr[pos] = new_val
        if current_l <= pos <= current_r:
            add_element(pos)
    
    def unapply_update(update, current_l, current_r):
        nonlocal arr
        pos, old_val, new_val = update
        if current_l <= pos <= current_r:
            remove_element(pos)
        arr[pos] = old_val
        if current_l <= pos <= current_r:
            add_element(pos)
    
    current_l = 0
    current_r = -1
    current_time = len(updates)
    ans_arr = [0] * len(queries)
    
    for query in queries_sorted:
        time_stamp, l, r, idx = query
        while current_time < time_stamp:
            apply_update(updates[current_time], current_l, current_r)
            current_time += 1
        while current_time > time_stamp:
            current_time -= 1
            unapply_update(updates[current_time], current_l, current_r)
        
        while current_r < r:
            current_r += 1
            add_element(current_r)
        while current_r > r:
            remove_element(current_r)
            current_r -= 1
        while current_l < l:
            remove_element(current_l)
            current_l += 1
        while current_l > l:
            current_l -= 1
            add_element(current_l)
        
        ans_arr[idx] = distinct_count - bad_count
    
    for ans in ans_arr:
        print(ans)

if __name__ == '__main__':
    main()